### ISSUE TITLE
Improve handling of MaybeIncorrect suggestions

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rustfix#200]
When the compiler or clippy produce suggestions, some of them are marked as MaybeIncorrect; i.e. the suggestion should probably not be _automatically_ applied. rustfix currently doesn't touch these.

This leads to confusion like in https://github.com/rust-lang/cargo/issues/8806, and is somewhat frustrating. We should _at the very least_ be clear that such suggestions were not applied.

_Ideally_, we should provide a mode where you can apply them anyway (`cargo clippy --fix --apply-all-suggestions`?), or apply them using a "pick and choose" mode.

See also: https://github.com/rust-lang/rustfix/issues/163

**Comment by Manishearth** [comment]
cc @estebank

**Comment by Manishearth** [comment]
also cc @ehuss for thinking about what the UX would be like from cargo's perspective

**Comment by estebank** [comment]
Is there something on the `rustc` side that needs to be done? We could add a new variant with a degree of confidence between `MachineApplicable` and `MaybeIncorrect`. We could have `LikelyIncorrect` for things we have little confidence in and start applying `MaybeIncorrect` on `rustfix`'s side.

**Comment by Manishearth** [comment]
@estebank no, I don't think that'll help. MaybeIncorrect already _is_ that variant: there's another one for when you're absolutely sure it's incorrect. MaybeIncorrect is just when it's a bad idea to auto apply the fix without checking; because the fix may introduce semantic change (etc)

At least the way it's used in clippy, most MaybeIncorrects are not supposed to be blindly applied.

cc @rust-lang/clippy in case folks have opinions on whether we need additional suggestion applicability types

This should be entirely on the rustfix/cargo side I think.

**Comment by matthiaskrgr** [comment]
I think to some extent rustfix already tries to skip applying lint results where the resulting code does not compile.
I would love to be able to just throw everything at it and let rustc and rustfix figure it out by themselves. :D

**Comment by Manishearth** [comment]
The problem is that MaybeIncorrect is often not stuff that is incorrect because it won't compile, it's incorrect because it changed the semantics of the program in a way clippy _thinks_ is according to intention but might not be

**Comment by flip1995** [comment]
> The problem is that MaybeIncorrect is often not stuff that is incorrect because it won't compile, it's incorrect because it changed the semantics of the program in a way clippy _thinks_ is according to intention but might not be

In Clippy `MaybeIncorrect` is used interchangeable  for both cases. Sometimes we use it, if the suggestion might result in new errors (e.g. the borrow checker might complain or an additional (trait) import is needed). And sometimes, because it's a semantic change (e.g. removes a possibly panicking expression). At least, that's how I use it and recommend it in reviews.

So maybe splitting it up in `MayIntroduceNewError` and `SemanticChange` (with possibly better names) would make sense.

**Comment by camsteffen** [comment]
Both reasons might apply. That may be reason enough not to split it. Unless we say, "if it is maybe incorrect for multiple reasons, prefer X." Is the motivation for splitting to offer an explanation to the user? Maybe instead we could add a "incorrect reason" field. In fact, in reviews I've picked up the habit of making sure there is a code comment explaining `MaybeIncorrect`.

**Comment by flip1995** [comment]
>  Maybe instead we could add a "incorrect reason" field.

That also sounds reasonable to me.

>  or apply them using a "pick and choose" mode

Especially this reason could be used in such a pick and choose prompt.

```
This suggestion was not automatically applied, because <reason>
Do you want to apply this suggestion anyway? (y/N)
```

**Comment by camsteffen** [comment]
For what it's worth, I think "may introduce error" can also be described as "maybe incomplete" - additional changes are necessary to "complete" the intended suggestion, like adding an import.

**Comment by jyn514** [comment]
> Ideally, we should provide a mode where you can apply them anyway (cargo clippy --fix --apply-all-suggestions?),

This exists today as `__CARGO_FIX_YOLO=1`. It would be great to make that more discoverable with a CLI option.

**Comment by NikosEfthias** [comment]
@jyn514 what does the YOLO stand for ?

**Comment by matthiaskrgr** [comment]
I think it tries to apply suggestions no matter what...?
https://github.com/rust-lang/cargo/blob/50a0af4bfd47294f53cf3000e1a7e076162280c6/src/cargo/ops/fix.rs#L581

**Comment by camsteffen** [comment]
All the kids are saying it

**Comment by jyn514** [comment]
@NikosEfthias sample usage:
> I was supposed to check if I accidentally commited tax crimes today via armchair lawyering  but we can just yolo that

**Comment by Manishearth** [comment]
@estebank @rust-lang/clippy it's been a while since it's been filed and there's some interest in exposing applicability levels to users. Before we do so, do y'all think it's worth splitting MaybeIncorrect into SemanticChange and MayIntroduceNewError?

**Comment by flip1995** [comment]
I think it does make sense to split that. We use this level for both of those use cases  interchangeably.

**Comment by Manishearth** [comment]
Maybe we can do a nice audit to clippy to see which goes where. I wonder if that's something @3tilley  is interested in  looking in to

The way I see it is that there are three somewhat independent work items here:

 - Look at migrating clippy and rustc over to a splti MaybeIncorrect level
 - Propose a cargo flag that exposes rustfix's yolo-mode. Design it better (perhaps as a "minimum applicability level" setting)
 - Add an interactive mode to rustfix.

**Comment by estebank** [comment]
Yeah, having that split makes sense to me. Migration is going to be annoying though.

**Comment by 3tilley** [comment]
Hello, I've been following this conversation but haven't had a chance to dig in properly. I YOLO'd this evening and it certainly did what it said on the tin. Just recording two and  a half bugs here while I've got them in front of me. This is git bash running on regular Windows 10 for reference:

1. The  `u` from `use` got dropped somehow:
![image](https://user-images.githubusercontent.com/1138504/205459437-a9e87f40-aa7d-4ab9-aa66-e4de46c17ed4.png)

2. Where use was required in multiple places, it was added multiple times
![image](https://user-images.githubusercontent.com/1138504/205459486-93d41ebd-daea-4a0f-b0db-77e1275a7d70.png)

3. I don't know if this is really a bug, but the format that was used for the `use` lines seems a bit odd - lots of blank lines added between statements. Are fixes supposed to be rustfmt compatible?

**Comment by llogiq** [comment]
I also agree with splitting `MaybeIncorrect`. Perhaps `MayFailToCompile`, `MayChangeBehavior` and `MayBeCounterproductive` (for things that might compile and keep behavior, but not help readability or perf).

**Comment by camsteffen** [comment]
How does splitting applicability improve UX? Or what problem does it solve?

**Comment by Manishearth** [comment]
@camsteffen It lets users set some minimum applicability level with rustfix, instead of going with the default.

The way I see it, the applicabilities go in order of least to most applicable:

 - HasPlaceholders
 - MaybeIncorrect (may change behavior)
 - MaybeIncorrect (may introduce error)
 - MachineApplicable

For example, quite often with the latter two I'm actually comfortable running rustfix on my codebase, with the former two I kinda want to review them case by case.

**Comment by kraktus** [comment]
As a clippy user I would also be interested in a better handling of `MaybeIncorrect` options. `clippy::semicolon_if_nothing_returned` is the canonical example of the lint that if failing sometimes automatically is not an issue, and on a large codebase it's much easier to dealt with only the corner cases manually than all the occurrences.

The split of `MaybeIncorrect` is probably worth being nominated in a clippy meeting?

**Comment by Qix-** [comment]
Just chiming in, being able to say something like `--try-anyway` even if it results in syntax errors etc. is still better than having to refactor 10,000 instances of something tedious (e.g. adding a generic to a deeply-coupled type).

**Comment by est31** [comment]
This would also be useful for rustc tests. It's not always the case that a `MaybeIncorrect` suggestion is incorrect, and you might want to add rustfix based tests for the cases where it is *right*, to ensure that there is no regression. e.g. I've filed https://github.com/rust-lang/rust/issues/105227 about an incorrect suggestion and the [PR to fix it](https://github.com/rust-lang/rust/pull/105805) would benefit from a rustfix option to apply the suggestion.

**Comment by ehuss** [comment]
@est31 compiletest applies all suggestions regardless of applicability with the `run-rustfix` header. Only when using `rustfix-only-machine-applicable` will it limit it to machine-applicable.

**Comment by est31** [comment]
@ehuss oh it seems compiletest seems to use `rustfix` as a crate and there you *can* tell it to apply `MaybeIncorrect` suggestions. Thanks for the pointer!

**Comment by Manishearth** [comment]
Here's the current plan:

- add PotentialBehaviorChange and PotentialCompileError. Keep MachineApplicable around for now. 
   - PotentialCompileError is considered more machine applicable than PotentialBehaviorChange since the latter means both
- do a campaign to migrate everyone
- remove MachineApplicable
- work with rustfix and RA for better UI


@veykril does this look OK to you from a rust-analyzer perspective? are these categories useful?

**Comment by ehuss** [comment]
I'm not sure I understand "remove MachineApplicable". Isn't that the variant that is most useful?  Did you maybe mean MaybeIncorrect?

Can you define exactly what these two categories mean? `MaybeIncorrect` is often used for suggestions that are known to be outright buggy. I'm not sure that really fits `PotentialCompileError`, which reads to me that `rustc` is unable to provide a better suggestion.

**Comment by Manishearth** [comment]
Er, sorry, meant MaybeIncorrect

> Can you define exactly what these two categories mean? `MaybeIncorrect` is often used for suggestions that are known to be outright buggy. I'm not sure that really fits `PotentialCompileError`, which reads to me that `rustc` is unable to provide a better suggestion.

 - PotentialCompileError: this may introduce a compiler error
 - PotentialBehaviorChange: this may change the behavior of the code, or introduce a compiler error (i.e., needs manual review)

It's an interesting point about suggestions that are outright buggy. Perhaps we should add a Nursery/Buggy category? Not sure. I kinda feel BehaviorChange covers that okayishly. Right now, MaybeIncorrect covers a lot of different types of incorrectness with lots of different implications on how people use the tooling. I'm not sure if Buggy and PotentialBehaviorChange has implications on tooling use

**Comment by estebank** [comment]
We might want to have an explicit `Applicability::Buggy` category. That would help rustfix to ignore them and is to audit them.

**Comment by jyn514** [comment]
> We might want to have an explicit `Applicability::Buggy` category. That would help rustfix to ignore them and is to audit them.

That's `Unspecified`, no?

https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/enum.Applicability.html

**Comment by Veykril** [comment]
> Here's the current plan:
> 
>     * add PotentialBehaviorChange and PotentialCompileError. Keep MaybeIncorrect around for now.
>       
>       * PotentialCompileError is considered more machine applicable than PotentialBehaviorChange since the latter means both
> 
>     * do a campaign to migrate everyone
> 
>     * remove MaybeIncorrect
> 
>     * work with rustfix and RA for better UI
> 
> 
> @Veykril does this look OK to you from a rust-analyzer perspective? are these categories useful?

I don't think rust-analyzer cares too much about this, I can't think of a direct use case for us here aside from maybe grouping quickfixes of these two kinds differently, as for both we'd want to show the user a quickfix in the UI (as we do with `MaybeIncorrect` in general right now).

**Comment by estebank** [comment]
@jyn514 it could be, I'm not sure if we are using it that way though.

**Comment by ehuss** [comment]
> That's `Unspecified`, no?

I think Unspecified was for diagnostics that have not been vetted as to whether or not they work correctly.  I believe that was used when the applicability system was put in, and there was a huge list of diagnostics to update, so Unspecified was the default until a more appropriate category could be determined.

**Comment by llogiq** [comment]
The idea behind splitting is that `MayNotCompile` suggestions *may* compile, and guarantee not to change behavior if they do. So a tool like rust fix could apply them, check compiling the result and roll them back on failure. On the other hand `MayChangeBehavior` have no such guarantee.

Changing the names will help us keep track of which suggestions have been vetted.

**Comment by Manishearth** [comment]
> I don't think rust-analyzer cares too much about this, I can't think of a direct use case for us here aside from maybe grouping quickfixes of these two kinds differently (r-a is doing a terrible job at categorizing quick fixes currently), as for both we'd want to show the user a quickfix in the UI (as we do with `MaybeIncorrect` in general right now).

Yeah I guess that r-a isn't batching things the way `rustfix` does so it's all user-driven anyway.




> The idea behind splitting is that `MayNotCompile` suggestions _may_ compile, and guarantee not to change behavior if they do. So a tool like rust fix could apply them, check compiling the result and roll them back on failure. On the other hand `MayChangeBehavior` have no such guarantee.

It also means that the user workflow can be "run rustfix with `PotentialCompileError`, and then manually fix any errors they see".

Whereas for `PotentialBehaviorChange` the workflow will probably be to review them one by one, and ideally put them in a separate commit for cleaner review (as a reviewer I'd want the direct rustfix fixes in one commit, `PotentialCompileError` fixups in another, and `PotentialBehaviorChange` ones in a third)

**Comment by flip1995** [comment]
> We might want to have an explicit `Applicability::Buggy` category.

`Unspecified` in Clippy is pretty much a shrug :shrug: for suggestions right now: no one spend time looking at the suggestion to categorize it. 

That said, I don't think a `Buggy` category would add much value here. I think, that every (potentially) "buggy" suggestion could be categorized into one of `PotentialCompileError`, `PotentialBehaviorChange`, or `HasPlaceholder`. And if it is really impossible to tell where it should go, I think `Unspecified` describes that pretty well.

**Comment by Manishearth** [comment]
cc @Muscraft
